src/Scriban/Syntax/ScriptVisitor.cs

```diff
@@ -0,0 +1,767 @@
+// Copyright (c) Alexandre Mutel. All rights reserved.
+// Licensed under the BSD-Clause 2 license.
+// See license.txt file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Scriban.Syntax
+{
+    public class ScriptVisitor
+    {
+        private readonly ScriptVisitorContext _context = new ScriptVisitorContext();
+
+        public IScriptVisitorContext Context => _context;
+
+        public ScriptNode Visit(ScriptNode node)
+        {
+            if (node == null) return null;
+
+            using (_context.Push(node))
```

Comment:
I know some people don't like this kind of "abuse" of `IDisposable`... Let me know if you prefer to change it to a try/finally instead.